default/python/AI/pathfinding.py

```diff
@@ -1,16 +1,137 @@
+             including both start and target system
+    :rtype: tuple[int]
+    """
+
+    universe = fo.getUniverse()
+    fleet = universe.getFleet(fleet_id)
+    if not fleet:
+        return None
+    empire = fo.getEmpire()
+    supplied_systems = set(empire.fleetSupplyableSystemIDs)
+    start_fuel = fleet.maxFuel if start in supplied_systems else fleet.fuel
+
+    # We have 1 free jump from supplied system into unsupplied systems.
+    # Thus, the target system must be at most maxFuel + 1 jumps away
+    # in order to reach the system under standard conditions.
+    # In some edge cases, we may have more supply here than what the
+    # supply graph suggests. For example, we could have recently refueled
+    # in a system that is now blockaded by an enemy or we have found
+    # the refuel special.
+    target_distance_from_supply = -min(state.get_system_supply(target), 0)
+    if (fleet.maxFuel + 1 < (target_distance_from_supply + minimum_fuel_at_target) and
+            universe.jumpDistance(start, target) > (start_fuel - minimum_fuel_at_target)):
+        # can't possibly reach this system with the required fuel
+        return None
+
+    mission_type = (mission_type_override if mission_type_override is not None else
+                    foAI.foAIstate.get_fleet_mission(fleet_id))
+    may_travel_starlane_func = _STARLANE_TRAVEL_FUNC_MAP.get(mission_type, _more_careful_travel_starlane_func)
+
+    path_info = find_path_with_resupply_generic(start, target, start_fuel, fleet.maxFuel,
+                                                lambda s: s in supplied_systems,
+                                                minimum_fuel_at_target,
+                                                may_travel_starlane_func=may_travel_starlane_func)
+
+    if _DEBUG_CHAT and may_travel_starlane_func != _always_ok_travel_starlane_func:
+        risky_path = find_path_with_resupply_generic(start, target, start_fuel, fleet.maxFuel,
+                                                     lambda s: s in supplied_systems,
+                                                     minimum_fuel_at_target)
+        if path_info and may_travel_starlane_func == _risky_travel_starlane_func:
+            safest_path = find_path_with_resupply_generic(start, target, start_fuel, fleet.maxFuel,
+                                                          lambda s: s in supplied_systems,
+                                                          minimum_fuel_at_target,
+                                                          may_travel_starlane_func=
+                                                                     _more_careful_travel_starlane_func)
+            if safest_path and path_info.distance < safest_path.distance:
+                message = "(Scout?) Fleet %d chose somewhat risky path %s instead of safe path %s" % (
+                    fleet_id, path_type, _info_string(path_info), _info_string(risky_path))
```

Comment:
This is broken: ```path_type``` is undeclared and there are 4 values for 3 placeholders in the string.

Reply:
Ah, right, I missed that one, taken care of now.